ipa-core/src/helpers/transport/mod.rs

```diff
@@ -333,6 +333,10 @@ pub trait Transport: Clone + Send + Sync + 'static {
     }
 }
 
+pub trait ShardedTransport: Transport {
```

Comment:
I think we should extend Transport trait instead with these 

```rust
    fn peer_count(&self) -> u32;

    /// Broadcasts a message to all peers, excluding this instance, collecting all failures and
    /// successes. This method waits for all responses and returns only when all peers responded.
    async fn broadcast<Q, S, R>(
        &self,
        route: R,
    ) -> Result<(), BroadcastError<Self::Identity, Self::Error>>
    where
        Option<QueryId>: From<Q>,
        Option<Gate>: From<S>,
        Q: QueryIdBinding,
        S: StepBinding,
        R: RouteParams<RouteId, Q, S> + Clone;
```

It is easy to implement, works for both MPC and Shard and avoids transport specialization which this code invites us to do

Reply:
I understand you would still allow Transport users to check the number of shards via `peer_count`?

Reply:
that's right. Depending on prepare query orchestration logic, you may just use `broadcast`